src/zkp/piprm.rs

```diff
@@ -156,42 +156,129 @@ impl Proof for PiPrmProof {
 
 #[cfg(test)]
 mod tests {
-    use crate::paillier::prime_gen;
-
     use super::*;
+    use crate::paillier::DecryptionKey;
+    use rand::Rng;
 
     fn random_ring_pedersen_proof<R: RngCore + CryptoRng>(
         rng: &mut R,
-    ) -> Result<(RingPedersen, PiPrmProof)> {
-        let (p, q) = prime_gen::get_prime_pair_from_pool_insecure(rng).unwrap();
-        let N = &p * &q;
-        let phi_n = (p - 1) * (q - 1);
-        let tau = BigNumber::from_rng(&N, rng);
-        let lambda = BigNumber::from_rng(&phi_n, rng);
-        let t = modpow(&tau, &BigNumber::from(2), &N);
-        let s = modpow(&t, &lambda, &N);
-
-        let ring_pedersen = RingPedersen::from_parts(s, t, N);
-        let proof = PiPrmProof::prove(rng, &ring_pedersen, &PiPrmSecret::new(lambda, phi_n))?;
-        Ok((ring_pedersen, proof))
+    ) -> Result<(RingPedersen, PiPrmProof, BigNumber, BigNumber)> {
+        let (sk, _, _) = DecryptionKey::new(rng)?;
+        let (scheme, lambda, totient) = RingPedersen::extract(&sk, rng)?;
+        let secrets = PiPrmSecret::new(lambda.clone(), totient.clone());
+        let proof = PiPrmProof::prove(rng, &scheme, &secrets)?;
+        Ok((scheme, proof, lambda, totient))
+    }
+
+    #[test]
+    fn piprm_proof_verifies() -> Result<()> {
+        let mut rng = crate::utils::get_test_rng();
+        let (input, proof, _, _) = random_ring_pedersen_proof(&mut rng)?;
+        proof.verify(&input)
+    }
+
+    #[test]
+    fn piprm_proof_serializes() -> Result<()> {
+        let mut rng = crate::utils::get_test_rng();
+        let (input, proof, _, _) = random_ring_pedersen_proof(&mut rng)?;
+        let serialized = bincode::serialize(&proof).unwrap();
+        let deserialized: PiPrmProof = bincode::deserialize(&serialized).unwrap();
+        assert_eq!(serialized, bincode::serialize(&deserialized).unwrap());
+        deserialized.verify(&input)
+    }
+
+    #[test]
+    fn incorrect_lengths_fails() -> Result<()> {
```

Comment:
suggestion: This only tests the too-short case. Can we also check too-long?

Reply:
Done!